### ISSUE TITLE
Loss function

### ISSUE THREAD

**Issue by enaserianhanzaei** [CuthbertCai/pytorch_DANN#13]
Hi,

I'm not sure if this is an issue, or that's my misunderstanding. You calculated loss as:
 
loss = class_loss + params.theta * domain_loss

shouldn't it be:

loss = class_loss - params.theta * domain_loss

according to the formula (9) of the paper?

**Comment by eyeris** [comment]
Hi I am having the same issue. Should be multi-lied by negative scaler. (Gradient Reversal Layer)

**Comment by vincent341** [comment]
> Hi,
> 
> I'm not sure if this is an issue, or that's my misunderstanding. You calculated loss as:
> 
> loss = class_loss + params.theta * domain_loss
> 
> shouldn't it be:
> 
> loss = class_loss - params.theta * domain_loss
> 
> according to the formula (9) of the paper?

Hi. I'm trying to understand the implementation and have the same question with you. Would you mind letting me know if you have known the answer? Much appreciated if I could hear from you.

**Comment by CuthbertCai** [comment]
> > Hi,
> > I'm not sure if this is an issue, or that's my misunderstanding. You calculated loss as:
> > loss = class_loss + params.theta * domain_loss
> > shouldn't it be:
> > loss = class_loss - params.theta * domain_loss
> > according to the formula (9) of the paper?
> 
> Hi. I'm trying to understand the implementation and have the same question with you. Would you mind letting me know if you have known the answer? Much appreciated if I could hear from you.

Maybe you can refer to `models.py`, there is a GradReverse Layer (GRL) to reverse gradients for layers before GRL.

**Comment by vincent341** [comment]
> > > Hi,
> > > I'm not sure if this is an issue, or that's my misunderstanding. You calculated loss as:
> > > loss = class_loss + params.theta * domain_loss
> > > shouldn't it be:
> > > loss = class_loss - params.theta * domain_loss
> > > according to the formula (9) of the paper?
> > 
> > 
> > Hi. I'm trying to understand the implementation and have the same question with you. Would you mind letting me know if you have known the answer? Much appreciated if I could hear from you.
> 
> Maybe you can refer to `models.py`, there is a GradReverse Layer (GRL) to reverse gradients for layers before GRL.

Thanks so much for the important hint. I understand now. Is removing [grad_versrse ](https://github.com/CuthbertCai/pytorch_DANN/blob/540e0c1699eece8496274342e561524429de9210/models/models.py#L83) layer and update the current loss function to _loss = class_loss - params.theta * domain_loss_  equivalent to the current implementation?

**Comment by CuthbertCai** [comment]
> > > > Hi,
> > > > I'm not sure if this is an issue, or that's my misunderstanding. You calculated loss as:
> > > > loss = class_loss + params.theta * domain_loss
> > > > shouldn't it be:
> > > > loss = class_loss - params.theta * domain_loss
> > > > according to the formula (9) of the paper?
> > > 
> > > 
> > > Hi. I'm trying to understand the implementation and have the same question with you. Would you mind letting me know if you have known the answer? Much appreciated if I could hear from you.
> > 
> > 
> > Maybe you can refer to `models.py`, there is a GradReverse Layer (GRL) to reverse gradients for layers before GRL.
> 
> Thanks so much for the important hint. I understand now. Is removing [grad_versrse ](https://github.com/CuthbertCai/pytorch_DANN/blob/540e0c1699eece8496274342e561524429de9210/models/models.py#L83) layer and update the current loss function to _loss = class_loss - params.theta * domain_loss_ equivalent to the current implementation?

I think the two implementations are not equivalent. Because loss function `loss=class_loss-params.theta*domain_loss` will reverse gradients of `extractor` and `domain_classifier` while GRL only reverses gradients of `extractor`.
